### PR TITLE
DTSPO-16054 Adding toplogySpreadConstraints option to pod sec

### PR DESCRIPTION
**Pull request by MartyFox**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DTSPO-16054

### Change description ###
Adding toplogySpreadConstraints option to pod sec.
If enabled it will allow users to spread workloads across a grouped set of nodes e.g. on-demand or spot instances.
This is currently set to use an input value but will likely get hardcoded at the base chart level dependent on the outcome of testing AKS Spot instances.

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### CODE REVIEW

File: library/templates/v2/_topologySpreadConstraints.tpl [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+{{- define "hmcts.topologySpreadConstraints.v2" }}
+{{- $languageValues := deepCopy .Values -}}
+{{- if hasKey .Values "language" -}}
+{{- $languageValues = (deepCopy .Values | merge (pluck .Values.language .Values | first) ) -}}
+{{- end -}}
+{{- if $languageValues.topologySpreadConstraints }}
+topologySpreadConstraints:
+  - maxSkew: {{ $languageValues.topologySpreadConstraints.maxSkew }}
+    topologyKey: {{ $languageValues.topologySpreadConstraints.topologyKey }}
+    whenUnsatisfiable: DoNotSchedule
```
Comment (timja) [inline library/templates/v2/_topologySpreadConstraints.tpl:14]: I expect we want to schedule anyway? what if there's no spot nodes available?
```suggestion
    whenUnsatisfiable: ScheduleAnyway
```

File: library/templates/v2/_topologySpreadConstraints.tpl [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    nodeTaintsPolicy: Honor
+    labelSelector:
+      matchLabels:
+        app.kubernetes.io/name: {{ template "hmcts.releasename.v2" . }}
+        helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    {{- if $languageValues.topologySpreadConstraints.minDomains }}
+    minDomains: {{ $languageValues.topologySpreadConstraints.minDomains }}
```
Comment (timja) [inline library/templates/v2/_topologySpreadConstraints.tpl:21]: if we change the `whenUnsatisfiable` then we can't use this anyway, still trying to understand what it does though.

File: library/templates/v2/_topologySpreadConstraints.tpl [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+{{/*
+Setup topologySpreadConstraints
+*/}}
+{{- define "hmcts.topologySpreadConstraints.v2" }}
+{{- $languageValues := deepCopy .Values -}}
+{{- if hasKey .Values "language" -}}
+{{- $languageValues = (deepCopy .Values | merge (pluck .Values.language .Values | first) ) -}}
+{{- end -}}
+{{- if $languageValues.topologySpreadConstraints }}
```
Comment (timja) [inline library/templates/v2/_topologySpreadConstraints.tpl:10]: I think we will want an `enabled` flag underneath `topologySpreadConstraints` that way we can set default values in the charts without requiring anyone to actually set these values and have it off by default

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 1 above)
Comment (adusumillipraveen) [reply library/templates/v2/_topologySpreadConstraints.tpl:14]: agree on ScheduleAnyway and make it configurable with default here.

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 3 above)
Comment (adusumillipraveen) [reply library/templates/v2/_topologySpreadConstraints.tpl:10]: yup, i am guessing we dont want to use this in prod ? so better have a global flag to enable/disable.

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 2 above)
Comment (MartyFox) [reply library/templates/v2/_topologySpreadConstraints.tpl:21]: If we change the value of `whenUnsatisfiable` then yes this becomes unnecessary and can be removed.
As it stands its a viable input however after testing may get removed as if we only plan to use for on-demand vs spot then our domains are set to 2 so theres no point in having the setting.

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 3 above)
Comment (MartyFox) [reply library/templates/v2/_topologySpreadConstraints.tpl:10]: Not sure what you mean here but happy to include it.

File: library/templates/v2/_topologySpreadConstraints.tpl [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+topologySpreadConstraints:
+  - maxSkew: {{ $languageValues.topologySpreadConstraints.maxSkew }}
+    topologyKey: {{ $languageValues.topologySpreadConstraints.topologyKey }}
+    whenUnsatisfiable: DoNotSchedule
+    nodeTaintsPolicy: Honor
+    labelSelector:
+      matchLabels:
+        app.kubernetes.io/name: {{ template "hmcts.releasename.v2" . }}
+        helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version }}
```
[the remark this answers is not in the document]
Comment (timja) [reply library/templates/v2/_topologySpreadConstraints.tpl:19]: teams bump their versions when they change the chart but the chart version you are getting there is the base chart not the teams chart version, i.e. it's not doing what you think and isn't working correctly. the base chart version will barely ever change.

and when new deployments are done the version won't change as they don't release a new chart version when they change the image.

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 3 above)
Comment (timja) [reply library/templates/v2/_topologySpreadConstraints.tpl:10]: something like this: (untested, done from memory)
```suggestion
{{- if $languageValues.topologySpreadConstraints and $languageValues.topologySpreadConstraints.enabled }}
```

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartyFox) [reply library/templates/v2/_topologySpreadConstraints.tpl:19]: Good to know, I'll remove it from here but there may be a need to have some other means of knowing the version or difference between deployments. Testing will hopefully prove that out.

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 3 above)
Comment (MartyFox) [reply library/templates/v2/_topologySpreadConstraints.tpl]: Global enable flag added

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartyFox) [reply library/templates/v2/_topologySpreadConstraints.tpl]: Chart label removed

File: library/templates/v2/_topologySpreadConstraints.tpl [hunk 5]
```diff
@@ -0,0 +1,23 @@
+---
+{{/*
+Setup topologySpreadConstraints
+*/}}
+{{- define "hmcts.topologySpreadConstraints.v2" }}
+{{- $languageValues := deepCopy .Values -}}
+{{- if hasKey .Values "language" -}}
+{{- $languageValues = (deepCopy .Values | merge (pluck .Values.language .Values | first) ) -}}
+{{- end -}}
+{{- if and $languageValues.topologySpreadConstraints $languageValues.global.topologySpreadConstraints.enabled }}
```
Comment (timja) [inline library/templates/v2/_topologySpreadConstraints.tpl:10]: why global? I don't think we need that for this.
```suggestion
{{- if and $languageValues.topologySpreadConstraints $languageValues.topologySpreadConstraints.enabled }}
```

File: ci-values-lang.yaml [hunk 6]
```diff
@@ -59,6 +61,9 @@ volumes:
       defaultMode: 0777
 nodeSelector:
   agentpool: cronjob
+topologySpreadConstraints:
```
Comment (timja) [inline ci-values-lang.yaml:64]:
```suggestion
topologySpreadConstraints:
  enabled: true
```

File: library/templates/v2/_topologySpreadConstraints.tpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (timja) [reply library/templates/v2/_topologySpreadConstraints.tpl]: I see you've solved it with https://github.com/hmcts/chart-library/pull/115/commits/eda6be8012c8e1af72da8b192fe441ade0b18aa7

**Review by timja — APPROVED** [review]

**Pull request merged**